crates/bevy_ecs/src/bundle/mod.rs

```diff
@@ -244,7 +244,13 @@ pub trait DynamicBundle {
     /// Calls `func` on each value, in the order of this bundle's [`Component`]s. This passes
     /// ownership of the component values to `func`.
     #[doc(hidden)]
-    fn get_components(self, func: &mut impl FnMut(StorageType, OwningPtr<'_>)) -> Self::Effect;
+    unsafe fn get_components(ptr: *const Self, func: &mut impl FnMut(StorageType, OwningPtr<'_>));
+
+    // SAFETY:
+    // - Must be called exactly once after `get_components` has been called.
+    // - `ptr` must point to a valid instance of `Self` but does not necessary need to be aligned.
+    #[doc(hidden)]
+    unsafe fn apply_effect(ptr: *const Self, entity: &mut EntityWorldMut);
```

Comment:
I'm not very fond of the `*const Self` API TBH, it sounds very easy to misuse

Reply:
This has been implemented following #20877.